## Exploratory Data Analysis

In [2]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import OrdinalEncoder,StandardScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from xgboost import XGBClassifier
import joblib
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, f1_score, recall_score, precision_score, classification_report, confusion_matrix, ConfusionMatrixDisplay
import warnings
warnings.filterwarnings("ignore")

In [10]:
df = pd.read_csv(r"C:\Users\DELL\Desktop\Backend\BACKEND-AI\MACHINE_LEARNING\Loan_Dataset\train.csv")
df.head(5)

id  annual_income  debt_to_income_ratio  credit_score  loan_amount  \
0   0       29367.99                 0.084           736      2528.42   
1   1       22108.02                 0.166           636      4593.10   
2   2       49566.20                 0.097           694     17005.15   
3   3       46858.25                 0.065           533      4682.48   
4   4       25496.70                 0.053           665     12184.43   

   interest_rate  gender marital_status education_level employment_status  \
0          13.67  Female         Single     High School     Self-employed   
1          12.92    Male        Married        Master's          Employed   
2           9.76    Male         Single     High School          Employed   
3          16.10  Female         Single     High School          Employed   
4          10.21    Male        Married     High School          Employed   

         loan_purpose grade_subgrade  loan_paid_back  
0               Other             C3             1.0  
1  Debt consolidation             D3             0.0  
2  Debt consolidation             C5             1.0  
3  Debt consolidation             F1             1.0  
4               Other             D1             1.0

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 593994 entries, 0 to 593993
Data columns (total 13 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   id                    593994 non-null  int64  
 1   annual_income         593994 non-null  float64
 2   debt_to_income_ratio  593994 non-null  float64
 3   credit_score          593994 non-null  int64  
 4   loan_amount           593994 non-null  float64
 5   interest_rate         593994 non-null  float64
 6   gender                593994 non-null  object 
 7   marital_status        593994 non-null  object 
 8   education_level       593994 non-null  object 
 9   employment_status     593994 non-null  object 
 10  loan_purpose          593994 non-null  object 
 11  grade_subgrade        593994 non-null  object 
 12  loan_paid_back        593994 non-null  float64
dtypes: float64(5), int64(2), object(6)
memory usage: 58.9+ MB


In [4]:
df.duplicated()

NameError: name 'df' is not defined

In [5]:
df.isna().sum()

NameError: name 'df' is not defined

In [6]:
df.describe().T

NameError: name 'df' is not defined

### Findings

From our dataset we noticed that we have

- 593994 entries, 13 columns
- The columns contains both numerical and categorical features
- All values are moderately close to the mean

### Recommendation 

From our finding we noticed that our data is clean and has no missing value in it. There is no prescence of outliers.

## Data Preprocessing

In [7]:
pd.DataFrame(df['loan_paid_back'].value_counts())

NameError: name 'df' is not defined

In [8]:
pd.DataFrame(df['grade_subgrade'].unique())


NameError: name 'df' is not defined

**Feature Engineering**

In [9]:
bins = [6000,	28000,	47000,	61000,	290000, float('inf')]
labels = ['Low', 'Medium', 'High', 'Very High', 'High earners']

# Create new column for income category
df['Income_Category'] = pd.cut(df['annual_income'], bins=bins, labels=labels)
df['Income_Category_Code'] = df['Income_Category'].map({'Low': 1, 'Medium': 2, 'High': 3, 'Very High': 4, 'High earners':5})

NameError: name 'df' is not defined

In [ ]:
income_category = pd.DataFrame(df["Income_Category"]).value_counts()
print(income_category)


Income_Category
Medium             156169
Low                149116
Very High          148176
High               140435
High earners           98
Name: count, dtype: int64


In [69]:
num_cols = ["Income_Category", "debt_to_income_ratio","credit_score","loan_amount","interest_rate","loan_paid_back"]
# for i in num_cols:
#         df[i] = np.log(df[i])

In [59]:
df[num_cols].dtypes


annual_income           float64
debt_to_income_ratio    float64
credit_score            float64
loan_amount             float64
interest_rate           float64
loan_paid_back          float64
dtype: object

In [1]:
num_df = df[num_cols]   
corr_df = pd.DataFrame({
    "Correlation Coefficient": np.abs(num_df.corrwith(num_df["loan_paid_back"]))
})

corr_df_sorted = corr_df.sort_values("Correlation Coefficient", ascending=False).iloc[1:11]

print(corr_df_sorted)


NameError: name 'df' is not defined

In [61]:
num_df = pd.DataFrame({"Correlation Coefficient": np.abs(num_cols.corrwith(num_cols["loan_paid_back"]))})
corr_df_sorted = num_df.sort_values("Correlation Coefficient", ascending=False).iloc[:, :10]
corr_df_sorted = corr_df_sorted.iloc[1:]
print(corr_df_sorted)

AttributeError: 'list' object has no attribute 'corrwith'